File: backend/app/funcional.py
"""
funcional.py
------------
Funciones PURAS (no modifican estado externo, mismo input -> mismo output).
Aquí se concentra la evidencia de PROGRAMACIÓN FUNCIONAL pedida en el
proyecto: funciones de orden superior, lambdas, map/filter/reduce.
"""

from functools import reduce
from typing import List, Dict, Any


# ---------- Cálculos de facturación (funciones puras) ----------

def calcular_subtotal(detalle: List[Dict[str, Any]]) -> float:
    """Suma cantidad*precio_unitario de cada línea usando map + reduce."""
    montos = map(lambda item: item["cantidad"] * item["precio_unitario"], detalle)
    return round(reduce(lambda acc, x: acc + x, montos, 0.0), 2)


def calcular_impuesto(subtotal: float, tasa: float = 0.15) -> float:
    return round(subtotal * tasa, 2)


def calcular_total(subtotal: float, impuesto: float) -> float:
    return round(subtotal + impuesto, 2)


def construir_detalle_factura(productos: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    productos: lista de dicts {producto, cantidad} donde 'producto' es un dict
    ya resuelto (con precio). Devuelve el detalle con subtotal por línea,
    usando map (función de orden superior).
    """
    def a_linea(item):
        p = item["producto"]
        cantidad = item["cantidad"]
        return {
            "producto_id": p["id"],
            "nombre": p["nombre"],
            "cantidad": cantidad,
            "precio_unitario": p["precio"],
            "subtotal": round(p["precio"] * cantidad, 2),
        }

    return list(map(a_linea, productos))


# ---------- Consultas / filtros sobre productos e inventario ----------

def filtrar_por_categoria(productos: List[Dict[str, Any]], categoria_id: int) -> List[Dict[str, Any]]:
    """Filtra productos por categoría usando filter + lambda."""
    return list(filter(lambda p: p["categoria_id"] == categoria_id, productos))


def productos_bajo_stock(productos: List[Dict[str, Any]], umbral: int = 5) -> List[Dict[str, Any]]:
    """Devuelve productos cuyo stock está por debajo del umbral (filter + lambda)."""
    return list(filter(lambda p: p["stock"] < umbral, productos))


def ordenar_por_precio(productos: List[Dict[str, Any]], descendente: bool = False) -> List[Dict[str, Any]]:
    """Orden funcional usando sorted + lambda (no muta la lista original)."""
    return sorted(productos, key=lambda p: p["precio"], reverse=descendente)


def valor_total_inventario(productos: List[Dict[str, Any]]) -> float:
    """Suma stock*precio de todos los productos usando map + reduce."""
    valores = map(lambda p: p["stock"] * p["precio"], productos)
    return round(reduce(lambda acc, v: acc + v, valores, 0.0), 2)


def resumen_inventario(productos: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Transforma productos a un resumen liviano usando map."""
    return list(map(
        lambda p: {
            "id": p["id"],
            "nombre": p["nombre"],
            "stock": p["stock"],
            "valor_en_stock": round(p["stock"] * p["precio"], 2),
        },
        productos,
    ))


def total_carrito(items_resueltos: List[Dict[str, Any]]) -> float:
    """items_resueltos: [{producto: {...precio}, cantidad: n}, ...]"""
    montos = map(lambda it: it["producto"]["precio"] * it["cantidad"], items_resueltos)
    return round(reduce(lambda acc, x: acc + x, montos, 0.0), 2)

File: tienda-cliente/app/assets/js/utils/funcional.js
import { TASA_IMPUESTO } from '../config.js';

export const calcularSubtotal = (items) => items.reduce((acumulado, item) => acumulado + item.subtotal, 0);

export const calcularImpuesto = (subtotal, tasa = TASA_IMPUESTO) => Number((subtotal * tasa).toFixed(2));

export const calcularTotal = (subtotal, impuesto) => Number((subtotal + impuesto).toFixed(2));

export const construirDetalleFactura = (items) =>
  items.map((item) => ({
    productoId: item.producto.id,
    nombre: item.producto.nombre,
    cantidad: item.cantidad,
    precioUnitario: item.producto.precio,
    subtotal: item.subtotal,
  }));

export const contarUnidades = (items) => items.reduce((total, item) => total + item.cantidad, 0);

export const filtrarPorCategoria = (productos, categoriaId) =>
  categoriaId === null || categoriaId === undefined || categoriaId === 'todas'
    ? productos
    : productos.filter((producto) => producto.categoriaId === Number(categoriaId));

export const buscarPorTexto = (productos, termino) => {
  const valor = termino.trim().toLowerCase();
  if (valor === '') return productos;
  return productos.filter(
    (producto) =>
      producto.nombre.toLowerCase().includes(valor) || producto.descripcion.toLowerCase().includes(valor)
  );
};

export const compararPor = (criterio, orden = 'asc') => (a, b) => {
  const factor = orden === 'asc' ? 1 : -1;
  if (a[criterio] < b[criterio]) return -1 * factor;
  if (a[criterio] > b[criterio]) return 1 * factor;
  return 0;
};

export const aplicarPipelineCatalogo = (productos, { categoriaId, termino, orden }) => {
  let resultado = filtrarPorCategoria(productos, categoriaId);
  resultado = buscarPorTexto(resultado, termino ?? '');
  if (orden) {
    resultado = [...resultado].sort(compararPor(orden.criterio, orden.direccion));
  }
  return resultado;
};
